File: app/src/main/java/uk/org/rivernile/android/bustracker/ui/search/SearchActivity.java
/*
 * Copyright (C) 2016 Niall 'Rivernile' Scott
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors or contributors be held liable for
 * any damages arising from the use of this software.
 *
 * The aforementioned copyright holder(s) hereby grant you a
 * non-transferrable right to use this software for any purpose (including
 * commercial applications), and to modify it and redistribute it, subject to
 * the following conditions:
 *
 *  1. This notice may not be removed or altered from any file it appears in.
 *
 *  2. Any modifications made to this software, except those defined in
 *     clause 3 of this agreement, must be released under this license, and
 *     the source code of any modifications must be made available on a
 *     publically accessible (and locateable) website, or sent to the
 *     original author of this software.
 *
 *  3. Software modifications that do not alter the functionality of the
 *     software but are simply adaptations to a specific environment are
 *     exempt from clause 2.
 */

package uk.org.rivernile.android.bustracker.ui.search;

import android.app.SearchManager;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.SearchRecentSuggestions;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

import uk.org.rivernile.android.bustracker.database.busstop.BusStopContract;
import uk.org.rivernile.android.bustracker.database.busstop.loaders.BusStopSearchLoader;
import uk.org.rivernile.android.bustracker.database.search.SearchSuggestionsProvider;
import uk.org.rivernile.android.bustracker.ui.bustimes.DisplayStopDataActivity;
import uk.org.rivernile.edinburghbustracker.android.R;

/**
 * This {@link android.app.Activity} allows the user to perform a search for bus stops.
 *
 * @author Niall Scott
 */
public class SearchActivity extends AppCompatActivity
        implements LoaderManager.LoaderCallbacks<Cursor>, SearchAdapter.OnItemClickedListener,
        SearchView.OnQueryTextListener, InstallBarcodeScannerDialogFragment.Callbacks {

    /**
     * If this {@link android.app.Activity} was started with
     * {@link #startActivityForResult(Intent, int)}, an {@link Intent} extra with this key will be
     * included in the result which contains the selected bus stop code.
     */
    public static final String EXTRA_STOP_CODE = "stopCode";

    private static final String BARCODE_APP_PACKAGE =
            "market://details?id=com.google.zxing.client.android";
    private static final String BARCODE_ACTION = "com.google.zxing.client.android.SCAN";
    private static final String BARCODE_EXTRA_QR_CODE_MODE = "QR_CODE_MODE";
    private static final String BARCODE_EXTRA_SCAN_RESULT = "SCAN_RESULT";

    private static final String DIALOG_INSTALL_QR_SCANNER = "installQrScanner";

    private static final int REQUEST_CODE_SCAN_QR = 1;

    private static final int LOADER_SEARCH = 1;

    private static final String LOADER_ARG_SEARCH_TERM = "searchTerm";

    private static final String URI_QUERY_PARAMETER_STOP_CODE = "busStopCode";

    private SearchRecentSuggestions recentSuggestions;
    private SearchAdapter adapter;

    private SearchView searchView;
    private RecyclerView recyclerView;
    private ProgressBar progress;
    private TextView txtError;

    private MenuItem menuItemScan;

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.search);
        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));

        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        searchView = (SearchView) findViewById(R.id.searchView);
        recyclerView = (RecyclerView) findViewById(android.R.id.list);
        progress = (ProgressBar) findViewById(R.id.progress);
        txtError = (TextView) findViewById(R.id.txtError);

        final SearchManager searchManager =
                (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setOnQueryTextListener(this);

        adapter = new SearchAdapter(this);
        adapter.setOnItemClickedListener(this);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(adapter);

        recentSuggestions = new SearchRecentSuggestions(this, SearchSuggestionsProvider.AUTHORITY,
                SearchSuggestionsProvider.MODE);

        showEmptySearchTermError();
        handleIntent(getIntent());
    }

    @Override
    protected void onNewIntent(final Intent intent) {
        super.onNewIntent(intent);

        setIntent(intent);
        handleIntent(intent);
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        super.onCreateOptionsMenu(menu);

        getMenuInflater().inflate(R.menu.search_option_menu, menu);
        menuItemScan = menu.findItem(R.id.search_option_menu_scan);

        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(final Menu menu) {
        super.onPrepareOptionsMenu(menu);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            menuItemScan.setVisible(getPackageManager()
                    .hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY));
        } else {
            menuItemScan.setVisible(getPackageManager()
                    .hasSystemFeature(PackageManager.FEATURE_CAMERA));
        }

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.search_option_menu_scan:
                handleQrCodeButtonClicked();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onActivityResult(final int requestCode, final int resultCode,
            final Intent data) {
        if (requestCode == REQUEST_CODE_SCAN_QR && resultCode == RESULT_OK) {
            // The data is delivered as an Intent, so if it's null the scan has not been successful,
            // despite the resultCode.
            if (data == null) {
                Toast.makeText(this, R.string.search_scan_error, Toast.LENGTH_SHORT).show();
                return;
            }

            // The scanned data is a URL, so parse it in to a Uri so we can easily extract the
            // stop code parameter.
            final Uri uri = Uri.parse(data.getStringExtra(BARCODE_EXTRA_SCAN_RESULT));

            // Only hierarchical URIs can possibly be valid, so this is a useful sanity check.
            if (!uri.isHierarchical()) {
                Toast.makeText(this, R.string.search_invalid_qrcode, Toast.LENGTH_SHORT).show();
                return;
            }

            final String stopCode = uri.getQueryParameter(URI_QUERY_PARAMETER_STOP_CODE);

            if (!TextUtils.isEmpty(stopCode)) {
                onStopCodeSelected(stopCode);
            } else {
                Toast.makeText(this, R.string.search_invalid_qrcode, Toast.LENGTH_SHORT).show();
            }
        }
    }

    @Override
    public Loader<Cursor> onCreateLoader(final int id, final Bundle args) {
        switch (id) {
            case LOADER_SEARCH:
                final String searchTerm =
                        args != null ? args.getString(LOADER_ARG_SEARCH_TERM) : null;

                return !TextUtils.isEmpty(searchTerm) ?
                        new BusStopSearchLoader(this, searchTerm) : null;
            default:
                return null;
        }
    }

    @Override
    public void onLoadFinished(final Loader<Cursor> loader, final Cursor data) {
        switch (loader.getId()) {
            case LOADER_SEARCH:
                handleSearchResultsLoaded(data);
                break;
        }
    }

    @Override
    public void onLoaderReset(final Loader<Cursor> loader) {
        switch (loader.getId()) {
            case LOADER_SEARCH:
                handleSearchResultsLoaded(null);
                break;
        }
    }

    @Override
    public void onItemClicked(@NonNull final Cursor cursor) {
        onStopCodeSelected(cursor.getString(
                cursor.getColumnIndex(BusStopContract.BusStops.STOP_CODE)));
    }

    @Override
    public boolean onQueryTextSubmit(final String query) {
        return false;
    }

    @Override
    public boolean onQueryTextChange(final String newText) {
        if (TextUtils.isEmpty(newText)) {
            showEmptySearchTermError();
            return true;
        } else {
            return false;
        }
    }

    @Override
    public void onShowInstallBarcodeScanner() {
        final Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(BARCODE_APP_PACKAGE));
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        try {
            startActivity(intent);
        } catch(ActivityNotFoundException e) {
            Toast.makeText(this, R.string.barcodescannerdialog_noplaystore, Toast.LENGTH_LONG)
                    .show();
        }
    }

    /**
     * Handle the {@link Intent} that this {@link android.app.Activity} was created with or a new
     * {@link Intent} that has since come in.
     *
     * @param intent The {@link Intent} to handle.
     */
    private void handleIntent(@NonNull final Intent intent) {
        if (!Intent.ACTION_SEARCH.equals(intent.getAction())) {
            return;
        }

        final String searchTerm = intent.getStringExtra(SearchManager.QUERY);
        searchView.setQuery(searchTerm, false);
        final Bundle loaderArgs = new Bundle();
        loaderArgs.putString(LOADER_ARG_SEARCH_TERM, searchTerm);

        final LoaderManager loaderManager = getSupportLoaderManager();
        final Loader<Cursor> loader = loaderManager.getLoader(LOADER_SEARCH);
        final BusStopSearchLoader searchLoader = (BusStopSearchLoader) loader;
        showProgress();

        if (loader != null && searchLoader.getSearchTerm().equals(searchTerm)) {
            loaderManager.initLoader(LOADER_SEARCH, loaderArgs, this);
        } else {
            recentSuggestions.saveRecentQuery(searchTerm, null);
            loaderManager.restartLoader(LOADER_SEARCH, loaderArgs, this);
        }
    }

    /**
     * Handle the result of loading search results.
     *
     * @param cursor The {@link Cursor} containing the search results.
     */
    private void handleSearchResultsLoaded(@Nullable final Cursor cursor) {
        adapter.swapCursor(cursor);

        if (cursor != null) {
            if (cursor.getCount() > 0) {
                showContent();
            } else {
                showNoResultsError();
            }
        } else {
            showEmptySearchTermError();
        }
    }

    /**
     * Show an error to the user when the search term is empty.
     */
    private void showEmptySearchTermError() {
        showError(R.string.search_error_empty);
    }

    /**
     * Show an error to the user when there were no search results.
     */
    private void showNoResultsError() {
        showError(R.string.search_error_no_results);
    }

    /**
     * Show a progress indicator to the user.
     */
    private void showProgress() {
        txtError.setVisibility(View.GONE);
        recyclerView.setVisibility(View.GONE);
        progress.setVisibility(View.VISIBLE);
    }

    /**
     * Show the content to the user, that is, the search results.
     */
    private void showContent() {
        txtError.setVisibility(View.GONE);
        progress.setVisibility(View.GONE);
        recyclerView.setVisibility(View.VISIBLE);
    }

    /**
     * Show an error to the user.
     *
     * @param errorResId The resource ID of the error string.
     */
    private void showError(@StringRes final int errorResId) {
        txtError.setText(errorResId);
        progress.setVisibility(View.GONE);
        recyclerView.setVisibility(View.GONE);
        txtError.setVisibility(View.VISIBLE);
    }

    /**
     * Handle a click on the scan QR code button.
     */
    private void handleQrCodeButtonClicked() {
        final Intent intent = new Intent(BARCODE_ACTION);
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
        intent.putExtra(BARCODE_EXTRA_QR_CODE_MODE, true);
        final List<ResolveInfo> packages = getPackageManager().queryIntentActivities(intent, 0);

        if (packages != null && !packages.isEmpty()) {
            try {
                startActivityForResult(intent, REQUEST_CODE_SCAN_QR);
            } catch (ActivityNotFoundException e) {
                showInstallQrScannerDialog();
            }
        } else {
            showInstallQrScannerDialog();
        }
    }

    /**
     * Show the dialog which asks the user if they wish to install a QR code scanning application.
     */
    private void showInstallQrScannerDialog() {
        new InstallBarcodeScannerDialogFragment()
                .show(getSupportFragmentManager(), DIALOG_INSTALL_QR_SCANNER);
    }

    /**
     * Handle a bus stop being selected by the user.
     *
     * @param stopCode The selected stop code.
     */
    private void onStopCodeSelected(@NonNull final String stopCode) {
        if (getCallingActivity() != null) {
            finishWithStopCode(stopCode);
        } else {
            showDisplayStopDetails(stopCode);
        }
    }

    /**
     * This is called if this {@link android.app.Activity} was started normally, i.e. not for a
     * result. This will start {@link DisplayStopDataActivity}.
     *
     * @param stopCode The selected stop code.
     */
    private void showDisplayStopDetails(@NonNull final String stopCode) {
        final Intent intent = new Intent(this, DisplayStopDataActivity.class);
        intent.putExtra(DisplayStopDataActivity.ARG_STOPCODE, stopCode);
        startActivity(intent);
    }

    /**
     * This is called if this {@link android.app.Activity} was started with
     * {@link #startActivityForResult(Intent, int)}. This will set the result and {@link #finish()}.
     *
     * @param stopCode The selected stop code.
     */
    private void finishWithStopCode(@NonNull final String stopCode) {
        final Intent result = new Intent();
        result.putExtra(EXTRA_STOP_CODE, stopCode);
        setResult(RESULT_OK, result);
        finish();
    }
}

File: app/src/main/kotlin/uk/org/rivernile/android/bustracker/ui/search/SearchActivity.kt
/*
 * Copyright (C) 2022 Niall 'Rivernile' Scott
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors or contributors be held liable for
 * any damages arising from the use of this software.
 *
 * The aforementioned copyright holder(s) hereby grant you a
 * non-transferrable right to use this software for any purpose (including
 * commercial applications), and to modify it and redistribute it, subject to
 * the following conditions:
 *
 *  1. This notice may not be removed or altered from any file it appears in.
 *
 *  2. Any modifications made to this software, except those defined in
 *     clause 3 of this agreement, must be released under this license, and
 *     the source code of any modifications must be made available on a
 *     publically accessible (and locateable) website, or sent to the
 *     original author of this software.
 *
 *  3. Software modifications that do not alter the functionality of the
 *     software but are simply adaptations to a specific environment are
 *     exempt from clause 2.
 *
 */

package uk.org.rivernile.android.bustracker.ui.search

import android.app.SearchManager
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.result.launch
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuProvider
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updateLayoutParams
import androidx.lifecycle.ViewModelProvider
import com.google.android.material.elevation.SurfaceColors
import com.google.android.material.shape.MaterialShapeDrawable
import dagger.android.AndroidInjection
import uk.org.rivernile.android.bustracker.core.text.TextFormattingUtils
import uk.org.rivernile.android.bustracker.map.StopMapMarkerDecorator
import uk.org.rivernile.android.bustracker.ui.bustimes.DisplayStopDataActivity
import uk.org.rivernile.edinburghbustracker.android.R
import uk.org.rivernile.edinburghbustracker.android.databinding.ActivitySearchBinding
import javax.inject.Inject

/**
 * This [android.app.Activity] allows the user to perform a text search for stops. Also, this is the
 * entry point to allow the user to perform a QR code scan for stops.
 *
 * @author Niall Scott
 */
class SearchActivity : AppCompatActivity(), InstallBarcodeScannerDialogFragment.Callbacks {

    companion object {

        /**
         * If this [android.app.Activity] was started with [startActivityForResult], an [Intent]
         * extra with this key will be included in the result which contains the selected stop code.
         */
        const val EXTRA_STOP_CODE = "stopCode"

        private const val DIALOG_INSTALL_QR_SCANNER = "installQrScannerDialog"

        private const val BARCODE_APP_PACKAGE =
                "market://details?id=com.google.zxing.client.android"
    }

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    @Inject
    lateinit var searchManager: SearchManager
    @Inject
    lateinit var stopMapMarkerDecorator: StopMapMarkerDecorator
    @Inject
    lateinit var textFormattingUtils: TextFormattingUtils

    private val viewModel: SearchActivityViewModel by viewModels { viewModelFactory }

    private lateinit var adapter: SearchAdapter

    private lateinit var viewBinding: ActivitySearchBinding

    private var menuItemScan: MenuItem? = null

    private val scanQrCodeLauncher = registerForActivityResult(ScanQrCode()) {
        viewModel.onQrScanned(it)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)

        super.onCreate(savedInstanceState)

        WindowCompat.setDecorFitsSystemWindows(window, false)
        viewBinding = ActivitySearchBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)

        setSupportActionBar(viewBinding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        viewBinding.appBarLayout.statusBarForeground =
                MaterialShapeDrawable.createWithElevationOverlay(this)
        window.navigationBarColor = SurfaceColors.SURFACE_2.getColor(this)

        ViewCompat.setOnApplyWindowInsetsListener(viewBinding.root) { view, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())

            view.updateLayoutParams<ViewGroup.MarginLayoutParams> {
                bottomMargin = insets.bottom
                leftMargin = insets.left
                rightMargin = insets.right
            }

            windowInsets
        }

        adapter = SearchAdapter(
                this,
                stopMapMarkerDecorator,
                textFormattingUtils,
                viewModel::onItemClicked)

        viewBinding.apply {
            recyclerView.apply {
                adapter = this@SearchActivity.adapter
                setHasFixedSize(true)
            }

            searchView.apply {
                setSearchableInfo(searchManager.getSearchableInfo(componentName))
                setOnQueryTextListener(queryTextListener)
            }
        }

        viewModel.searchResultsLiveData.observe(this, adapter::submitList)
        viewModel.uiStateLiveData.observe(this, this::handleUiStateChanged)
        viewModel.isScanMenuItemVisibleLiveData.observe(this, this::handleIsScanMenuItemVisible)
        viewModel.showStopLiveData.observe(this, this::handleShowStop)
        viewModel.showQrCodeScannerLiveData.observe(this) {
            handleShowQrCodeScanner()
        }
        viewModel.showInstallQrScannerDialogLiveData.observe(this) {
            handleShowInstallQrScannerDialog()
        }
        viewModel.showInvalidQrCodeErrorLiveData.observe(this) {
            showInvalidQrCodeError()
        }

        addMenuProvider(menuProvider)
        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        this.intent = intent
        handleIntent(intent)
    }

    override fun onShowInstallBarcodeScanner() {
        try {
            Intent(Intent.ACTION_VIEW)
                    .setData(Uri.parse(BARCODE_APP_PACKAGE))
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    .let(this::startActivity)
        } catch (e: ActivityNotFoundException) {
            Toast.makeText(
                    this,
                    R.string.barcodescannerdialog_noplaystore,
                    Toast.LENGTH_LONG)
                    .show()
        }
    }

    /**
     * Handle a new [Intent] for this [android.app.Activity].
     *
     * @param intent The new [Intent].
     */
    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH != intent.action) {
            return
        }

        val searchTerm = intent.getStringExtra(SearchManager.QUERY)
        viewBinding.searchView.setQuery(searchTerm, false)
        viewModel.searchTerm = searchTerm
    }

    /**
     * Handle the UI state changing.
     *
     * @param state The new [UiState].
     */
    private fun handleUiStateChanged(state: UiState) {
        viewBinding.apply {
            when (state) {
                is UiState.EmptySearchTerm -> {
                    txtError.apply {
                        setText(R.string.search_error_empty)
                        setCompoundDrawablesWithIntrinsicBounds(0, R.drawable.ic_error_search, 0, 0)
                    }

                    contentView.showErrorLayout()
                }
                is UiState.InProgress -> contentView.showProgressLayout()
                is UiState.NoResults -> {
                    txtError.apply {
                        setText(R.string.search_error_no_results)
                        setCompoundDrawablesWithIntrinsicBounds(
                                0, R.drawable.ic_error_directions_bus, 0, 0)
                    }

                    contentView.showErrorLayout()
                }
                is UiState.Content -> contentView.showContentLayout()
            }
        }
    }

    /**
     * Handle the visibility of the scan QR code menu item changing.
     *
     * @param isVisible Is the scan QR code menu item visible?
     */
    private fun handleIsScanMenuItemVisible(isVisible: Boolean) {
        menuItemScan?.isVisible = isVisible
    }

    /**
     * Handle how to show the stop. If this [android.app.Activity] was started by another Activity
     * which wishes to obtain a result, then we call [finishWithStopCode], otherwise we call
     * [showDisplayStopDetails].
     *
     * @param stopCode The selected stop code.
     */
    private fun handleShowStop(stopCode: String) {
        callingActivity?.let {
            finishWithStopCode(stopCode)
        } ?: showDisplayStopDetails(stopCode)
    }

    /**
     * Launch [DisplayStopDataActivity] with the selected [stopCode].
     *
     * @param stopCode The stop code to launch [DisplayStopDataActivity] with.
     */
    private fun showDisplayStopDetails(stopCode: String) {
        Intent(this, DisplayStopDataActivity::class.java)
                .putExtra(DisplayStopDataActivity.EXTRA_STOP_CODE, stopCode)
                .let(this::startActivity)
    }

    /**
     * When this [android.app.Activity] has been started by another Activity and we're finishing
     * (because either a valid QR code was scanned, or the user selected a search result item),
     * then set the result and then [finish] ourselves.
     *
     * @param stopCode The selected stop code.
     */
    private fun finishWithStopCode(stopCode: String) {
        val intent = Intent().putExtra(EXTRA_STOP_CODE, stopCode)
        setResult(RESULT_OK, intent)
        finish()
    }

    /**
     * Attempt to launch the QR code scanner application.
     */
    private fun handleShowQrCodeScanner() {
        try {
            scanQrCodeLauncher.launch()
        } catch (ignored: ActivityNotFoundException) {
            viewModel.onQrScannerNotFound()
        }
    }

    /**
     * Show a dialog to the user asking them if they wish to install the QR scanner application.
     */
    private fun handleShowInstallQrScannerDialog() {
        InstallBarcodeScannerDialogFragment()
                .show(supportFragmentManager, DIALOG_INSTALL_QR_SCANNER)
    }

    /**
     * Show a [Toast] notification which informs the user they scanned an invalid QR code.
     */
    private fun showInvalidQrCodeError() {
        Toast.makeText(this, R.string.search_invalid_qrcode, Toast.LENGTH_SHORT).show()
    }

    private val queryTextListener = object : SearchView.OnQueryTextListener {

        override fun onQueryTextSubmit(query: String?): Boolean {
            viewModel.submitSearchTerm(query)
            return true
        }

        override fun onQueryTextChange(newText: String?): Boolean {
            viewModel.searchTerm = newText
            return false
        }
    }

    private val menuProvider = object : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            menuInflater.inflate(R.menu.search_option_menu, menu)
            menuItemScan = menu.findItem(R.id.search_option_menu_scan)
        }

        override fun onPrepareMenu(menu: Menu) {
            handleIsScanMenuItemVisible(viewModel.isScanMenuItemVisibleLiveData.value ?: false)
        }

        override fun onMenuItemSelected(menuItem: MenuItem) = when (menuItem.itemId) {
            R.id.search_option_menu_scan -> {
                viewModel.onScanMenuItemClicked()
                true
            }
            else -> false
        }
    }
}